Introduce PrivateBPF= to use BPF tokens

teknoraver opened the pull request:
Add a new option `PrivateBPF=` to allow usage of **BPF** tokens within a
namespace.
`PrivateBPF=` can be set to `no` to use the host bpffs, `yes` to do a new
mount, and `token` to use the BPF token API: https://lwn.net/Articles/947173/

Currently not working because to create a BPF token the user namespace must be unshared first, and the mount namespace later.
Systemd is doing the opposite, so a refactor is needed to have it working.

DaanDeMeyer reviewed (changes requested):
Can we split this up into one commit that adds PrivateBPF=yes/no just for mounting bpffs and another commit that adds PrivateBPF=token?

Also after taking a more detailed look I think wedging this into setup_private_users() is the wrong way to go. Instead, let's implement this in setup_private_bpf() as a separate function.

The other issue with the current approach is that you're unsharing namespaces as part of the bpf token logic. This is not the way to go, systemd already has its own namespace unsharing logic and that should be reused instead of unsharing namespaces as part of setting up the token.

What this means is that we'll need two functions: prepare_bpf_token() and finalize_bpf_token(). All that prepare_bpf_token() should do is fork and then run the code from sd_bpffs() if the token is enabled, nothing more. finalize_bpf_token() should run after all other namespace setup is completed and all it should do is fsopen("bpf"), send it to the socket from the privileged helper process, receive the bpffs mount fd from the privileged helper process socket, create the token from it and then mount it.

More specifically, neither prepare_bpf_token() nor finalize_bpf_token() should unshare any namespaces. prepare_bpf_token() will have to return the write end of the socketpair which should then be passed to finalize_bpf_token() so it can communicate with the privileged helper process spawned in prepare_bpf_token().

DaanDeMeyer commented on src/core/namespace.c:
```diff
@@ -823,6 +834,19 @@ static int append_protect_system(MountList *ml, ProtectSystem protect_system, bo
+        case PRIVATE_BPF_YES:
+                return append_static_mounts(ml, private_bpf_yes_table, ELEMENTSOF(private_bpf_yes_table), ignore_protect);
+        default:
+                return -EINVAL;
```
`assert_not_reached()`

DaanDeMeyer commented on meson.build:
```diff
@@ -1639,6 +1639,10 @@ foreach term : ['analyze',
         conf.set10(name, have)
 endforeach
 
+libbpf = dependency('libbpf', required : false)
```
We already do this lookup earlier in this file, please move this stuff there

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5423,6 +5462,36 @@ int exec_invoke(
+                if (r < 0)
+                        return r;
+
+                r = wait_for_terminate_and_check("(bpffs)", TAKE_PID(bpftoken_pid), 0);
+                if (r < 0)
+                        return r;
+                if (r != EXIT_SUCCESS) /* If something strange happened with the child, let's consider this fatal, too */
+                        return -EIO;
```
Should be moved before receive_one_fd(), otherwise send_one_fd() may stuck on failure.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5423,6 +5462,36 @@ int exec_invoke(
+                if (mnt_fd < 0)
+                        return fs_fd;
+
+                bpffs_fd = openat(mnt_fd, ".", O_RDONLY);
+                if (bpffs_fd < 0)
+                        return bpffs_fd;
```
Why is it necessary to open??

teknoraver replied:
It was present in the sample code, but maybe it's unneeded now that I'm not creating the token anymore:

https://elixir.bootlin.com/linux/v6.14.6/source/tools/testing/selftests/bpf/prog_tests/token.c#L374

teknoraver commented on src/core/load-fragment.c:
```diff
@@ -126,6 +126,10 @@ int parse_confirm_spawn(const char *value, char **console) {
         return 0;
 }
 
+static int parse_delegate(const char *s) {
+        return -1;
+}
```
yes, I'm writing a real parser with 4 different config options: BPFDelegate{Commands,Maps,Programs,Attachments}=

teknoraver replied:
@yuwata I've added a draft parser for one of these 4 commands, but when I set BPFDelegateCommands=XX I don't see the value in bpffs_helper, it's always 0.

Can you check what is wrong before I start writing the other 3 commands parser?

teknoraver commented on src/core/exec-invoke.c:
```diff
@@ -2233,6 +2233,75 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+                        report_errno_and_exit(errno_pipe[1], r);
+                }
+
+                /* If the file is missing the kernel is too old, let's continue anyway. */
```
I don't know, I just moved that code from below to there.
But I could investigate and add that comment anyway.

teknoraver replied:
BTW It was introduced in 3.19:

https://github.com/torvalds/linux/commit/9cc46516ddf497ea16e8d7cb986ae03a0f6b92f8

yuwata replied:
Then, maybe we can make the condition more strict?

teknoraver commented:
@yuwata I finally have the `BPFDelegateCommands` correctly propagated to exec-invoke.c

The other three commands will be a copy-paste of this one, can you review this one so I will implement the others after this one is acceptable?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,6 +2244,29 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+        if (r < 0)
+                _exit(EXIT_FAILURE);
+
+        exit(EXIT_SUCCESS);
```
`exit()` -> `_exit()`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5435,6 +5473,32 @@ int exec_invoke(
+        if (context->private_bpf) {
+                _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
+
+                fs_fd = fsopen("bpf", 0);
+                if (fs_fd < 0)
+                        return fs_fd;
+
+                r = send_one_fd(token_fds[0], fs_fd, 0);
+                if (r < 0)
+                        return r;
```
I am not familiar, but why fsopen() needs to be called in the parent process?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5334,6 +5360,18 @@ int exec_invoke(
+                if (r < 0)
+                        return r;
+
+                r = safe_fork("(bpffs)", FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, &bpftoken_pid);
```
For newer calls, maybe better to use PidRef version?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5334,6 +5360,18 @@ int exec_invoke(
+
+                r = safe_fork("(bpffs)", FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, &bpftoken_pid);
+                if (r < 0)
+                        return r;
```
Please set `exit_status` on failure.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5435,6 +5473,32 @@ int exec_invoke(
+
+                r = move_mount(mnt_fd, "", AT_FDCWD, "/sys/fs/bpf", MOVE_MOUNT_F_EMPTY_PATH);
+                if (r < 0)
+                        return r;
```
Please set `exit_status` on failure in this block.

yuwata commented on src/core/namespace.c:
```diff
@@ -919,6 +924,19 @@ static int append_protect_system(MountList *ml, ProtectSystem protect_system, bo
+
+        switch (private_bpf) {
+        case PRIVATE_BPF_NO:
+                return append_static_mounts(ml, private_bpf_no_table, ELEMENTSOF(private_bpf_no_table), ignore_protect);
```
Not addressed?

teknoraver replied:
Sorry, I forgot this one while focusing on BPFDelegateCommands

yuwata commented on src/core/dbus-util.h:
```diff
@@ -179,7 +179,7 @@ int bus_property_get_triggered_unit(sd_bus *bus, const char *path, const char *i
                                                                         \
                 const char *s;                                          \
-                type v;                                                 \
+                type v = (type)0;                                       \
```
Why necessary to initialize? If it is necessary, why the cast is necessary?

At least, please insert a space after the cast.

teknoraver replied:
Because GCC reports a false warning and bails out with an error:
```
In file included from ../src/core/dbus-execute.c:15:
In function ‘bus_set_transient_bpf_delegate_commands’,
    inlined from ‘bus_exec_context_set_transient_property’ at ../src/core/dbus-execute.c:2212:24:
../src/core/dbus-util.h:197:28: error: ‘v’ may be used uninitialized [-Werror=maybe-uninitialized]
  197 |                         *p = v;                                         \
      |                         ~~~^~~
../src/core/dbus-execute.c:1681:8: note: in expansion of macro ‘BUS_DEFINE_SET_TRANSIENT_PARSE_PTR’
 1681 | static BUS_DEFINE_SET_TRANSIENT_PARSE_PTR(bpf_delegate_commands, uint64_t, bpf_delegate_from_string);
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../src/core/dbus-execute.c: In function ‘bus_exec_context_set_transient_property’:
../src/core/dbus-util.h:182:22: note: ‘v’ declared here
  182 |                 type v;                                                 \
      |                      ^
../src/core/dbus-execute.c:1681:8: note: in expansion of macro ‘BUS_DEFINE_SET_TRANSIENT_PARSE_PTR’
 1681 | static BUS_DEFINE_SET_TRANSIENT_PARSE_PTR(bpf_delegate_commands, uint64_t, bpf_delegate_from_string);
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

yuwata replied:
Then, please add short comments about that. You can find such several places.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,14 +2244,42 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
 }
 
-_noreturn_ static void bpffs_helper(int parent_fd) {
+_noreturn_ static void bpffs_helper(int parent_fd,
```
Please break line after `(`, and indent arguments with 16ch.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,14 +2244,42 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+                                    uint64_t delegate_maps,
+                                    uint64_t delegate_progs,
+                                    uint64_t delegate_attachs) {
         _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
```
Please insert an empty line after arguments before variable declarations.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,14 +2244,42 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+                                    uint64_t delegate_progs,
+                                    uint64_t delegate_attachs) {
         _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
+        char number[16 + 2 + 1];
```
Please make this more descriptive, e.g.
`STRLEN("0x") + sizeof(uint64_t) * 2 + 1` ?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,14 +2244,42 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
         if (fs_fd < 0)
                 _exit(EXIT_FAILURE);
 
+        if (delegate_cmds != ~0ULL) {
```
`!= UINT64_MAX` ?

teknoraver replied:
As you prefer, that was how the corresponding kernel code generates that number:

https://elixir.bootlin.com/linux/v6.14.8/source/kernel/bpf/inode.c#L917

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2244,14 +2244,42 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
                 _exit(EXIT_FAILURE);
 
+        if (delegate_cmds != ~0ULL) {
+                sprintf(number, "0x%lx", delegate_cmds);
```
Please use `xsprintf()` and `PRIx64`

BTW, why not to use `bpf_delegate_to_string()` here?

yuwata commented on src/core/load-fragment.c:
```diff
@@ -6645,3 +6646,27 @@ int config_parse_protect_hostname(
+        assert(rvalue);
+
+        if (streq(rvalue, "any")) {
+                *sz = ~0ULL;
```
UINT64_MAX

BTW, I think it is better to use `bpf_delegate_from_string()` here.

yuwata commented on src/core/namespace.c:
```diff
@@ -3855,6 +3856,27 @@ int refresh_extensions_in_namespace(
 }
 
+void bpf_delegate_to_string(uint64_t u, char *s) {
+        if (u == ~0ULL)
+                strcpy(s, "any");
+        else
+                sprintf(s, "0x%lx", u);
```
Here also, please use `UINT64_MAX` and `PRIx64`.

yuwata commented on src/core/namespace.h:
```diff
@@ -234,6 +234,9 @@ ProcSubset proc_subset_from_string(const char *s) _pure_;
 const char* private_bpf_to_string(PrivateBPF i) _const_;
 PrivateBPF private_bpf_from_string(const char *s) _pure_;
 
+void bpf_delegate_to_string(uint64_t u, char *s);
```
How about something like the following?
```c
#define BPF_DELEGATE_STRING_MAX (STRLEN("0x") + sizeof(uint64_t) * 2 + 1)
char* bpf_delegate_to_string(uint64_t u, char buf[static BPF_DELEGATE_STRING_MAX]);
#define BPF_DELEGATE_TO_STRING(u) bpf_delegate_to_string(u, (char[BPF_DELEGATE_STRING_MAX]){})
```
Then, you can use e.g.
```c
r = serialize_item(f, "exec-context-bpf-delegate-commands", BPF_DELEGATE_TO_STRING(c->bpf_delegate_commands));
``

teknoraver replied:
is the assert still needed with buf declared as array?

yuwata commented on src/core/namespace.c:
```diff
@@ -3855,6 +3856,27 @@ int refresh_extensions_in_namespace(
+int bpf_delegate_from_string(const char *s, uint64_t *u) {
+        int r;
+
+        r = safe_atou64(s, u);
```
Maybe better to use `safe_atoux64()` ?

yuwata commented on src/core/namespace.c:
```diff
@@ -3855,6 +3856,27 @@ int refresh_extensions_in_namespace(
+
+        r = safe_atou64(s, u);
+        if (r < 0) {
+                if (streq(s, "any"))
+                        *u = ~0ULL;
```
Not a matter, but I'd like to check with `"any"` first, then call `safe_atoux64()` (or `safe_atou64`).

yuwata commented on src/core/namespace.c:
```diff
@@ -3855,6 +3856,27 @@ int refresh_extensions_in_namespace(
+                sprintf(s, "0x%lx", u);
+}
+
+int bpf_delegate_from_string(const char *s, uint64_t *u) {
```
Missing assertion for `s`.
Please rename `u` -> `ret`

yuwata commented on src/core/namespace.h:
```diff
@@ -234,6 +234,9 @@ ProcSubset proc_subset_from_string(const char *s) _pure_;
 PrivateBPF private_bpf_from_string(const char *s) _pure_;
 
+void bpf_delegate_to_string(uint64_t u, char *s);
+int bpf_delegate_from_string(const char *s, uint64_t *u) _pure_;
```
`u` -> `ret`

yuwata commented on src/core/namespace.c:
```diff
@@ -3855,6 +3856,27 @@ int refresh_extensions_in_namespace(
         return 0;
 }
 
+void bpf_delegate_to_string(uint64_t u, char *s) {
```
missing assertion for `s`

yuwata commented on src/core/execute.c:
```diff
@@ -1134,7 +1135,8 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
                 prefix, proc_subset_to_string(c->proc_subset),
-                prefix, private_bpf_to_string(c->private_bpf));
+                prefix, private_bpf_to_string(c->private_bpf),
+                prefix, c->bpf_delegate_commands);
```
`%s` with `BPF_DELEGATE_TO_STRING(c->bpf_delegate_commands)` should be better?

teknoraver replied:
but bpf_delegate_to_string() wants a preallocated buffer, it's not worth it

teknoraver replied:
ah I just saw your proposal below

yuwata commented on src/core/execute-serialize.c:
```diff
@@ -2756,6 +2764,8 @@ static int exec_context_deserialize(ExecContext *c, FILE *f) {
                         if (c->private_bpf < 0)
                                 return -EINVAL;
+                } else if ((val = startswith(l, "exec-context-bpf-delegate-commands="))) {
+                        (void) bpf_delegate_from_string(val, &c->bpf_delegate_commands);
```
Why voidified? Please propagate error code.

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -4638,6 +4663,8 @@ int exec_invoke(
         int ngids = 0, ngids_after_pam = 0;
         int socket_fd = -EBADF, named_iofds[3] = EBADF_TRIPLET;
         size_t n_storage_fds, n_socket_fds, n_extra_fds;
+        _cleanup_close_pair_ int token_fds[2] = EBADF_PAIR;
+        _cleanup_(pidref_done_sigkill_wait) PidRef bpftoken_pid = PIDREF_NULL;
```
We don't do anything with bpf token, so these aren't appropriate variable names to use.

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5330,6 +5357,20 @@ int exec_invoke(
+                if (r < 0)
+                        return r;
+
+                r = pidref_safe_fork("(bpffs)", FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, &bpftoken_pid);
```
sd-bpffs

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5330,6 +5357,20 @@ int exec_invoke(
+                        return r;
+                }
+                if (r == 0)
+                        bpffs_helper(token_fds[1]);
```
Please move all of this into bpffs_helper() and return the pidref and socket as output arguments.

And rename `bpffs_helper()` to `bpffs_prepare()` or something like that.

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5431,6 +5472,45 @@ int exec_invoke(
         if (r < 0)
                 return r;
 
+        if (context->private_bpf) {
+                _cleanup_close_ int fs_fd = -EBADF, mnt_fd = -EBADF;
+
+                fs_fd = fsopen("bpf", 0);
+                if (fs_fd < 0) {
+                        *exit_status = EX_OSERR;
+                        return fs_fd;
+                }
+
+                r = send_one_fd(token_fds[0], fs_fd, 0);
+                if (r < 0) {
+                        *exit_status = EX_IOERR;
+                        return r;
+                }
+
+                r = wait_for_terminate_and_check("(bpffs)", TAKE_PIDREF(bpftoken_pid).pid, 0);
+                if (r < 0) {
+                        *exit_status = EX_OSERR;
+                        return r;
+                }
+                if (r != EXIT_SUCCESS) {
+                        /* If something strange happened with the child, let's consider this fatal, too */
+                        *exit_status = EX_OSERR;
+                        return -EIO;
+                }
+
+                mnt_fd = receive_one_fd(token_fds[0], 0);
+                if (mnt_fd < 0) {
+                        *exit_status = EX_IOERR;
+                        return fs_fd;
+                }
```
All the above can be done after the first call to setup_private_users(). Please move the logic into a separate function that is called after setup_private_users().

teknoraver replied:
but setup_private_users() is done before setup_delegated_namespaces(), the fsopen() call must be done after all the unshares

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5431,6 +5472,45 @@ int exec_invoke(
+                        return fs_fd;
+                }
+
+                r = move_mount(mnt_fd, "", AT_FDCWD, "/sys/fs/bpf", MOVE_MOUNT_F_EMPTY_PATH);
+                if (r < 0) {
+                        *exit_status = EX_OSERR;
+                        return r;
+                }
```
This should still be done in setup_namespace(), you can pass the fd as part of `NamespaceParameters` to make it available in `setup_namespace()`

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2244,6 +2244,72 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
         return 0;
 }
 
+_noreturn_ static void sd_userns(int errno_pipe[2], int unshare_ready_fd,
```
please give it a proper name. setup_private_users_child?

DaanDeMeyer replied:
Tbh I'd just drop this commit, there's no point in splitting it out anymore now that we've changed the approach.

teknoraver replied:
It helps to keep setup_private_users() smaller.
It's a static inline, what's wrong with keeping the split anyway?

yuwata replied:
Yeah, I like the change.

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2348,72 +2414,8 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
-                _exit(EXIT_SUCCESS);
-        }
+        if (r == 0)
+                sd_userns(errno_pipe, unshare_ready_fd, uid_map, gid_map, allow_setgroups);
```
A more idiomatic pattern is to make the child a "normal" function, i.e. returns negative errno on error, and then call _exit/report_errno_and_exit accordingly here.

YHNdnzj replied:
See for instance run_editor and run_editor_child

yuwata commented on src/core/load-fragment.c:
```diff
@@ -6645,3 +6646,20 @@ int config_parse_protect_hostname(
         return 1;
 }
+
+int config_parse_bpf_delegate(
```
DEFINE_CONFIG_PARSE_PTR(...) can be used for declaring such simple conf parser.

teknoraver commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -3340,6 +3340,8 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
       readonly (ss) ProtectHostnameEx = ...;
       @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
+      readonly s PrivateBPF = '...';
+      @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
```
where, in `<title>History</title>`?

yuwata replied:
Yes.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
 
+static int bpffs_prepare(
+                int *parent_fd,
+                PidRef *bpffs_pid) {
```
Please rename
- `parent_fd` -> `ret_pipe_fd`
- `bpffs_pid` -> `ret_pid`

At least, please prefix two arguments with `ret_`.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+                int *parent_fd,
+                PidRef *bpffs_pid) {
+
+        _cleanup_close_pair_ int bpffs_fds[2] = EBADF_PAIR;
```
Also here, naming is spurious. How about simply `pipe_fds`?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+
+        assert(parent_fd);
+
+        r = socketpair(AF_UNIX, SOCK_SEQPACKET, 0, bpffs_fds);
```
missing `SOCK_CLOEXEC`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+        return 0;
+}
+
+static int bpffs_finalize(int bpffs_fd, int pid) {
```
Missing assertions for `bpffs_fd`, also rename to `pipe_fd` or so.
Please use `PidRef` rather than `int pid`.

Also I'd like to rename `bpffs_fd` -> `pipe_fd` or so to indicate it is a socket.

teknoraver replied:
and put `TAKE_PIDREF(ret_pid)` after the call to bpffs_finalize()

yuwata replied:
Sorry, I cannot follow?

teknoraver replied:
sorry, I commented under the wrong place

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+        if (r < 0)
+                return r;
+
+        r = wait_for_terminate_and_check("(sd-bpffs)", pid, /* flags = */ 0);
```
Then, please use `pidref_wait_for_terminate_and_check()`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+        if (mnt_fd < 0)
+                return fs_fd;
+
+        r = move_mount(mnt_fd, "", AT_FDCWD, "/sys/fs/bpf", MOVE_MOUNT_F_EMPTY_PATH);
+        if (r < 0)
+                return -errno;
+
+        return 0;
```
Not a matter, but
```c
return RET_NERRNO(move_mount(...));
```

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -4640,8 +4715,9 @@ int exec_invoke(
         _cleanup_free_ gid_t *gids = NULL, *gids_after_pam = NULL;
         int ngids = 0, ngids_after_pam = 0;
-        int socket_fd = -EBADF, named_iofds[3] = EBADF_TRIPLET;
+        int socket_fd = -EBADF, named_iofds[3] = EBADF_TRIPLET, bpffs_fd = -EBADF;
```
bpffs_fd -> bpffs_socket_fd is better?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5441,6 +5525,14 @@ int exec_invoke(
                 return r;
 
+        if (context->private_bpf) {
+                r = bpffs_finalize(bpffs_fd, TAKE_PIDREF(bpffs_pid).pid);
```
If `bpffs_finalize()` fails earlier, the child process may not be sigkilled.

teknoraver replied:
because of TAKE_PIDREF()?

yuwata replied:
Right.

yuwata commented on src/core/namespace.c:
```diff
@@ -2653,6 +2679,10 @@ int setup_namespace(const NamespaceParameters *p, char **reterr_path) {
         if (r < 0)
                 return r;
 
+        r = append_private_bpf(&ml, p->private_bpf, p->protect_kernel_tunables, false);
```
`/* ignore_protect = */ false`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5340,6 +5416,14 @@ int exec_invoke(
                 }
         }
 
+        if (context->private_bpf) {
```
`context->private_bpf` is not a boolean.
```c
if (context->private_bpf != PRIVATE_BPF_NO) {
```

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5441,6 +5525,14 @@ int exec_invoke(
         if (r < 0)
                 return r;
 
+        if (context->private_bpf) {
```
Also here.
```c
if (context->private_bpf != PRIVATE_BPF_NO) {
```

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5340,6 +5416,14 @@ int exec_invoke(
+                r = bpffs_prepare(&bpffs_fd, &bpffs_pid);
+                if (r < 0) {
+                        *exit_status = EXIT_BPF;
+                        return log_error_errno(r, "Failed to mount BPFFS: %m");
```
`BPFFS` -> `bpffs` ?
The error message is inconsistent with the one for `bpffs_finalize()`.

yuwata commented on src/core/namespace.c:
```diff
@@ -343,7 +347,8 @@ static bool mount_entry_read_only(const MountEntry *p) {
 
-        return p->noexec || IN_SET(p->mode, MOUNT_NOEXEC, MOUNT_INACCESSIBLE, MOUNT_PRIVATE_SYSFS, MOUNT_BIND_SYSFS, MOUNT_PROCFS, MOUNT_PRIVATE_CGROUP2FS);
+        return p->noexec || IN_SET(p->mode, MOUNT_NOEXEC, MOUNT_INACCESSIBLE, MOUNT_PRIVATE_SYSFS,
+                        MOUNT_BIND_SYSFS, MOUNT_PROCFS, MOUNT_PRIVATE_CGROUP2FS);
```
Please drop the irrelevant change from the second commit.

yuwata commented on tools/dbus_ignorelist:
```diff
@@ -904,6 +904,7 @@ org.freedesktop.systemd1.Mount.Options
 org.freedesktop.systemd1.Mount.PAMName
 org.freedesktop.systemd1.Mount.PassEnvironment
 org.freedesktop.systemd1.Mount.Personality
+org.freedesktop.systemd1.Mount.PrivateBPF
```
Why the property is added to the ignore list?
Instead, please mention the property in History section.

teknoraver replied:
I saw other similar properties like `PrivateTmp`, `ProtectProc` being added too.
What's the logic behind the ignore list?

yuwata commented on man/systemd.exec.xml:
```diff
@@ -2525,6 +2525,18 @@ RestrictNamespaces=~cgroup net</programlisting>
+
+        <listitem><para>Accepts a bitmask in hexadecimal representing the allowed BPF commands, or "any"
+        to allow everything. Requires <varname>PrivateBPF=</varname> enabled to be effective.
+        This will be set the <constant>delegate_cmds</constant> bpffs mount option.
```
`will be set` -> `will set` ?

yuwata commented on man/systemd.exec.xml:
```diff
@@ -2525,6 +2525,18 @@ RestrictNamespaces=~cgroup net</programlisting>
+        to allow everything. Requires <varname>PrivateBPF=</varname> enabled to be effective.
+        This will be set the <constant>delegate_cmds</constant> bpffs mount option.
+        A more detailed explanation of the feature can be found in this
+        <ulink url="https://lwn.net/Articles/947173/">LWN post</ulink></para>
```
Missing period at the end of the sentence.

Please also mention the default value. Defaults to 0 ? or `"any"` ?

yuwata commented on man/systemd.exec.xml:
```diff
@@ -2525,6 +2525,18 @@ RestrictNamespaces=~cgroup net</programlisting>
+      <varlistentry>
+        <term><varname>BPFDelegateCommands=</varname></term>
+
+        <listitem><para>Accepts a bitmask in hexadecimal representing the allowed BPF commands, or "any"
```
`"any"` -> `<literal>any</literal>`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2243,6 +2248,24 @@ static int bpffs_prepare(
 
+                bpf_delegate_to_string(delegate_cmds, number);
+
+                r = fsconfig(fs_fd, FSCONFIG_SET_STRING, "delegate_cmds", number, 0);
```
`number` -> `BPF_DELEGATE_TO_STRING(delegate_cmds)`

teknoraver replied:
The kernel API accepts a numerical bitmask.
When you print the mount options the bitmask is pretty printed into a list of tokens, but when you set the options, the kernel accepts only a number.

yuwata commented on src/core/execute-serialize.c:
```diff
@@ -1836,6 +1836,13 @@ static int exec_context_serialize(const ExecContext *c, FILE *f) {
         if (r < 0)
                 return r;
 
+        if (c->bpf_delegate_commands) {
```
`c->bpf_delegate_commands != 0`

yuwata reviewed (changes requested):
Looks mostly good, at least superficially.

BTW, maybe this is still WIP, as delegate_maps and friends are unconditionally "any", and relevant settings are ignored.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -5441,6 +5534,15 @@ int exec_invoke(
+                        *exit_status = EXIT_BPF;
+                        return log_error_errno(r, "Failed to mount bpffs: %m");
+                }
+                TAKE_PIDREF(bpffs_pid);
```
TAKE_PIDREF() should be moved to after pidref_wait_for_terminate_and_check() in bpffs_finalize().

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2236,45 +2236,61 @@ static int bpffs_prepare(
         r = pidref_safe_fork("(sd-bpffs)", FORK_RESET_SIGNALS|FORK_DEATHSIG_SIGKILL, ret_pid);
         if (r < 0)
-                return r;
+                return log_error_errno(errno, "Failed to fork: %m");
```
`errno` -> `r`

teknoraver replied:
Hi, I'm adding logs here just to debug why PrivateBPF is working locally and failing in the CI.
But if you want I can keep them, with better error messages.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2236,45 +2236,61 @@ static int bpffs_prepare(
                 r = fsconfig(fs_fd, FSCONFIG_SET_STRING, "delegate_cmds", BPF_DELEGATE_TO_STRING(delegate_cmds), 0);
-                if (r < 0)
+                if (r < 0) {
+                        log_error_errno(r, "Failed to fsconfig FSCONFIG_SET_STRING: %m");
```
`r` -> `errno`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2236,45 +2236,61 @@ static int bpffs_prepare(
                 r = fsconfig(fs_fd, FSCONFIG_SET_STRING, "delegate_maps", "any", 0);
-                if (r < 0)
+                if (r < 0) {
+                        log_error_errno(r, "Failed to fsconfig FSCONFIG_SET_STRING: %m");
```
Also here and below.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2236,45 +2236,61 @@ static int bpffs_prepare(
                 mnt_fd = fsmount(fs_fd, 0, 0);
-                if (mnt_fd < 0)
+                if (mnt_fd < 0) {
+                        log_error_errno(mnt_fd, "Failed to receive_one_fd: %m");
```
`mnt_fd` -> `errno`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2236,45 +2236,61 @@ static int bpffs_prepare(
                 r = send_one_fd(pipe_fds[0], mnt_fd, 0);
-                if (r < 0)
+                if (r < 0) {
+                        log_error_errno(errno, "Failed to send_one_fd: %m");
```
`errno` -> `r`

yuwata reviewed (changes requested):
Maybe better to introduce error pipe and use `report_errno_and_exit()` on failure.

teknoraver commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -3342,6 +3342,8 @@ node /org/freedesktop/systemd1/unit/avahi_2ddaemon_2eservice {
       @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
       readonly s PrivateBPF = '...';
       @org.freedesktop.DBus.Property.EmitsChangedSignal("const")
+      readonly t BPFDelegateCommands = ...;
```
Yes, that's what the kernel does when showing mountpoints:
```
bash-5.2# mount |grep -w bpf
none on /sys/fs/bpf type bpf (rw,relatime,delegate_cmds=map_create:map_lookup_elem:map_update_elem:map_get_next_key,delegate_maps=any,delegate_progs=any,delegate_attachs=any)
```

I can expand the bitmask into tokens later

teknoraver replied:
Syntax will be like:

```
# systemd-run -p PrivateUsers=yes -p PrivateMounts=yes -p DelegateNamespaces=mnt -p PrivateBPF=yes -p BPFDelegateCommands=BPFMapCreate,BPFMapLookupElem,BPFProgRun -S
Running as unit: run-p2230-i2530.service; invocation ID: 705076f90dfe421cb9de85e1047f1621

bash-5.2# mount |grep -w bpf
none on /sys/fs/bpf type bpf (rw,relatime,delegate_cmds=map_create:map_lookup_elem:prog_test_run,delegate_maps=any,delegate_progs=any,delegate_attachs=any)
```

yuwata commented:
(Setting this PR draft, as it is not completed. Feel free to push ready for review button.)

teknoraver commented:
@yuwata thanks, I couldn't mark it as draft, maybe it's some github setting

teknoraver commented on src/core/exec-invoke.c:
```diff
@@ -2219,6 +2220,79 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+        assert(pipe_fd >= 0);
+        assert(pid);
+
+        fs_fd = fsopen("bpf", 0);
```
I have done this in a draft commit marked as [TEST].
I will squash it later

teknoraver commented:
@DaanDeMeyer @yuwata PR is ready to review, last needed thing is the script to generate the documentation from bpf.h

yuwata commented:
[Lint Code Base / Lint Code Base](https://github.com/systemd/systemd/actions/runs/15881239875/job/44781856076?pr=36134) CI is red.

yuwata commented on src/core/namespace.c:
```diff
@@ -927,6 +966,42 @@ static int append_protect_system(MountList *ml, ProtectSystem protect_system, bo
+                r = pidref_wait_for_terminate_and_check("(sd-bpffs)", p->bpffs_pid, /* flags = */ 0);
+                TAKE_PIDREF(*p->bpffs_pid);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to pidref_wait_for_terminate_and_check: %m");
```
the log message is slightly borked?
Maybe something `Failed to wait for subprocess configuring bpffs` ?

BTW, this is a debugging log, so not a big matter, but pidref_wait_for_terminate_and_check() logs all failures. Hence the logging here and below is not necessary.

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1772,24 @@ static int mount_overlay(const MountEntry *m) {
+
+        r = move_mount(mnt_fd, "", AT_FDCWD, mount_entry_path(m), MOVE_MOUNT_F_EMPTY_PATH);
+        if (r < 0)
+                return log_debug_errno(errno, "Failed to move_mount: %m");
```
`"Failed to move bpffs mount to %s: %m", mount_entry_path(m)` ?

teknoraver replied:
or maybe `m->path_const` which is the destination path without the prefix?

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1772,24 @@ static int mount_overlay(const MountEntry *m) {
 }
 
+static int mount_bpffs(const MountEntry *m) {
+        assert(m);
```
Please add an assertion for `m->bpffs_fd`.

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1772,24 @@ static int mount_overlay(const MountEntry *m) {
+static int mount_bpffs(const MountEntry *m) {
+        assert(m);
+
+        _cleanup_close_ int mnt_fd = fsmount(m->bpffs_fd, 0, 0);
+        int r;
+
+        safe_close(m->bpffs_fd);
+
+        if (mnt_fd < 0)
+                log_debug_errno(errno, "Failed to fsmount: %m");
```
Missing `return` on failure. Please improve the debugging log.
First declare `int r;`, then assertions, then fsmount():
```c
static int mount_bpffs(const MountEntry *m) {
        int r;

        assert(m);
        assert(m->bpffs_fd >= 0);

        _cleanup_close_ int mnt_fd = fsmount(m->bpffs_fd, 0, 0);
        if (mnt_fd < 0)
                return log_debug_errno(errno, "Failed to fsmount bpffs: %m");
```

`m->bpffs_fd` should not be closed here. If `bpffs_finalize()` succeed, but setting mount namespace failed before this function is called, the file descriptor is leaked.

yuwata commented on src/core/namespace.c:
```diff
@@ -2540,6 +2637,20 @@ int setup_namespace(const NamespaceParameters *p, char **reterr_path) {
                         return r;
         }
 
+        if (unshare(CLONE_NEWNS) < 0) {
```
I do not think this should be moved.
Rather than move this, how about to call bpffs_finalize() and friends in mount_bpffs() ?

teknoraver replied:
done, in this way I don't leak the fs anymore

yuwata commented on src/core/namespace.c:
```diff
@@ -101,6 +102,7 @@ typedef struct MountEntry {
         bool noexec:1;            /* Shall set MS_NOEXEC on the mount itself */
         bool exec:1;              /* Shall clear MS_NOEXEC on the mount itself */
         bool create_source_dir:1; /* Create the source directory if it doesn't exist - for implicit bind mounts */
+        bool bpffs:1;             /* Entry is a BPF filesystem, and has to be mounted with fsmount() and move_mount() */
```
The flag is unused?

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2328,69 +2375,10 @@ static int setup_private_users(PrivateUsers private_users, uid_t ouid, gid_t ogi
-                        r = log_debug_errno(errno, "Failed to write UID map to %s: %m", a);
+                r = setup_private_users_child(unshare_ready_fd, uid_map, gid_map, allow_setgroups);
+                if (r < 0)
                         report_errno_and_exit(errno_pipe[1], r);
```
Not your fault, but the reported error code is unused?

teknoraver replied:
Right. We could fix it in another PR, fixing it here is too much at once.

yuwata replied:
Yeah, I agree.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2270,6 +2271,50 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+                fs_fd = receive_one_fd(pipe_fds[0], 0);
+                if (fs_fd < 0) {
+                        log_debug_errno(fs_fd, "Failed to receive_one_fd from parent: %m");
+                        _exit(EXIT_FAILURE);
```
I'd like to also introduce error_pipe and use `report_errno_and_exit()` here.

yuwata commented on man/systemd.exec.xml:
```diff
@@ -2525,6 +2525,19 @@ RestrictNamespaces=~cgroup net</programlisting>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>BPFDelegateCommands=</varname></term>
```
The commit message is outdated.

yuwata commented on man/systemd.exec.xml:
```diff
@@ -2525,6 +2525,19 @@ RestrictNamespaces=~cgroup net</programlisting>
+      <varlistentry>
+        <term><varname>BPFDelegateCommands=</varname></term>
+
+        <listitem><para>Accepts a list of BPF commands to delegate or <literal>any</literal> to allow everything,
```
`any` -> `all` is better?
Please mention the list of possible values of BPF commands.

teknoraver replied:
"any" is consinstent with the kernel API, which uses "any" to delegate all types

yuwata replied:
OK.

yuwata commented on src/basic/generate-bpf-delegate-configs.py:
```diff
@@ -0,0 +1,69 @@
+import sys
+
+def print_usage_and_exit():
+    print(f'Usage: {sys.argv[0]} code <header>')
```
Not a matter, but `code` -> `header` is better? Not a strong opinion, feel free to ignore.

yuwata commented on src/basic/meson.build:
```diff
@@ -283,8 +283,19 @@ filesystem_switch_case_inc = custom_target(
+fname = 'bpf-delegate-configs.inc'
+bpf_delegate_configs_inc = custom_target(
+        fname,
+        input : 'include/linux/bpf.h',
```
Please use `files('include/linux/bpf.h')`

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2290,16 +2291,45 @@ static int bpffs_prepare(
                 r = fsconfig(fs_fd, FSCONFIG_CMD_CREATE, NULL, NULL, 0);
-                if (r < 0)
+                if (r < 0) {
+                        log_debug_errno(errno, "Failed to FSCONFIG_CMD_CREATE: %m");
```
The debug logging is inserted in a wrong commit.

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(private_bpf, PrivateBPF);
 
+#include "bpf-delegate-configs.inc"
```
Uh, please use `.h` suffix.

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                fprintf(fd, "any");
+        else {
+                bool first = true;
+                for (uint64_t i = 0; i < sizeof(u) * 8; i++) {
+                        if (u & (1UL << i)) {
```
Please use BIT_FOREACH()

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+        _cleanup_free_ char *buf = NULL;
+        size_t len = 1;
+
+        for (uint64_t i = 0; i < sizeof(u) * 8; i++)
+                if (u & (1UL << i))
```
Also here: `BIT_FOREACH()`

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                if (u & (1UL << i))
+                        len += strlen(bpf_delegate_cmd_table[i]) + 1;
+
+        buf = malloc(len);
```
I do not think it is necessary to optimize here.
Please use `strextendf_with_separator()`

yuwata replied:
Hm, but we already do like that at many places..
Rather than optimize here, I think it is better to improve `strextend()` and friends. Then we can speed up many places.
Maybe, caching the current string length should be enough, like we do in strv_extend_n().

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+
+DEFINE_STRING_TABLE_LOOKUP(bpf_delegate_cmd, uint64_t);
+
+int bpf_dump_delegate_commands(FILE *fd, uint64_t u) {
```
FILE is not a fd. Please rename to `f`.

BTW, I do not think it is necessary to introduce similar functions.
Please drop this and always use bpf_delegate_commands_to_string().

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                return 0;
+        }
+
+        _cleanup_strv_free_ char **tokens = strv_split(s, ",");
+
+        *ret = 0;
+
+        for (char **token = tokens; *token; token++) {
```
Please use extract_first_word() in the loop.

teknoraver replied:
To strip the whitespaces?

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                if (i < 0)
+                        return log_error_errno(i, "Invalid BPF delegate command %s", *token);
+
+                *ret |= 1UL << i;
```
- `1UL` -> `UINT64_C(1)`
- Please do not update `*ret` partially, as the next loop may fail. Please store the result to tentative variable, then assign it to `ret` at the end of the function.

yuwata commented on src/test/test-bpf-token.c:
```diff
@@ -0,0 +1,21 @@
+
+#include "tests.h"
+
+int main(void)
+{
```
Please move `{` to the end of the previous line.

yuwata commented on src/test/test-bpf-token.c:
```diff
@@ -0,0 +1,21 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <fcntl.h>
+#include <bpf/bpf.h>
```
Please sort.

yuwata commented on src/test/test-bpf-token.c:
```diff
@@ -0,0 +1,21 @@
+        int bpffs_fd = open("/sys/fs/bpf", O_RDONLY);
+        assert_se(bpffs_fd >= 0);
+
+        int token_fd = bpf_token_create(bpffs_fd, NULL);
```
Not a big matter, but this tests no our code... What do you want to test?

At least please split the test case to TEST(bpf_token_create), and use DEFINE_TEST_MAIN() to declare the main function.

Also, maybe better to check capability. Otherwise, if this runs on a container without CAP_BPF, it may fail. Though, I've not checked. So, may not necessary.

Moreover, please test bpf_delegate_commands_to_string() and friends here (by moving them somewhere, maybe in src/shared/bpf-token-util.c or so).

teknoraver replied:
This is not a unit test, but a helper used by the real test which is in TEST-07-PID1.
This is why I set it to "manual" so it's not run by meson.

This binary ensures that the bpffs is mounted correctly: calling it with a regular bpffs mount fails.

yuwata commented on src/core/namespace.c:
```diff
@@ -3996,6 +3997,84 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                return 0;
+        }
+
+        _cleanup_strv_free_ char **tokens = strv_split(s, ",");
```
At least, missing error handling.

yuwata commented on src/core/dbus-util.h:
```diff
@@ -179,7 +179,9 @@ int bus_property_get_triggered_unit(sd_bus *bus, const char *path, const char *i
-                type v;                                                 \
+                /* Need to initialize this to avoid a false positive    \
+                 * ‘v’ may be used uninitialized [-Werror=maybe-uninitialized] */ \
+                type v = 0;                                             \
```
This should not be false positive. But real bug in the parser.

teknoraver replied:
So what, just edit the comment?

yuwata replied:
No no, I mean, it is a bug in your new parser.

teknoraver commented on test/integration-tests/integration-test-wrapper.py:
```diff
@@ -617,6 +617,7 @@ def main() -> None:
         '--credential', f"journal.storage={'persistent' if sys.stdin.isatty() else args.storage}",
         *(['--runtime-build-sources=no', '--register=no'] if not sys.stdin.isatty() else []),
         'vm' if args.vm or os.getuid() != 0 or os.getenv('TEST_PREFER_QEMU', '0') == '1' else 'boot',
+        *(['--', '--capability=CAP_BPF'] if not args.vm and os.getenv('TEST_PREFER_QEMU', '0') == '0' else []),
```
To separate the mkosi arguments to the nspawn ones

yuwata replied:
Hm, I did not know that. Thanks.

teknoraver commented on src/core/exec-invoke.c:
```diff
@@ -5408,6 +5488,27 @@ int exec_invoke(
+                 * This is the kernel sample doing this:
+                 * https://github.com/torvalds/linux/blob/master/tools/testing/selftests/bpf/prog_tests/token.c
+                 */
+                r = pipe2(error_pipe, O_CLOEXEC|O_NONBLOCK);
+                if (r < 0)
+                        return log_debug_errno(errno, "Failed to create pipe: %m");
```
Which pipe is opened in bpffs_prepare().
I can only find `pipe_fds` which, despite the name, is a socket pair used by the fsopen API to pass the filesystem file descriptor between the privileged and unprivileged processes via SCM_RIGHTS.

teknoraver commented:
@yuwata @DaanDeMeyer I've addressed all the comments and implemented the remaining three commands, with an extra unit test. Also the CI passed.
Now I would like to fix the generation of the HTML doc with `generate-bpf-delegate-configs.py`

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1732,35 @@ static int mount_overlay(const MountEntry *m) {
+
+        r = send_one_fd(socket_fd, fs_fd, /* flags = */ 0);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to send_one_fd to child: %m");
```
`send_one_fd` -> `send bpffs fd`

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1732,35 @@ static int mount_overlay(const MountEntry *m) {
+
+        _cleanup_close_ int bpffs_fd = receive_one_fd(socket_fd, /* flags = */ 0);
+        if (bpffs_fd < 0)
+                return log_debug_errno(bpffs_fd, "Failed to receive_one_fd from child: %m");
```
`receive_one_fd` -> `receive bpffs fd`

yuwata commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -12089,7 +12185,11 @@ $ gdbus introspect --system --dest org.freedesktop.systemd1 \
+      <varname>BPFDelegateCommands</varname>,
+      <varname>BPFDelegateMaps</varname>,
+      <varname>BPFDelegatePrograms</varname>,
+      <varname>BPFDelegateAttachments</varname>, and
```
Also here. Also add them to socket, mount, and swap units.

yuwata commented on src/core/execute.c:
```diff
@@ -1156,6 +1160,29 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
 
+        if (c->private_bpf == PRIVATE_BPF_YES) {
+                _cleanup_free_ char *commands = bpf_delegate_commands_to_string(c->bpf_delegate_commands);
+                if (!commands)
+                        return (void) log_oom();
```
This is only dumping. Please ignore the failures.

yuwata commented on src/core/execute.c:
```diff
@@ -1156,6 +1160,29 @@ void exec_context_dump(const ExecContext *c, FILE* f, const char *prefix) {
+                if (!attachments)
+                        return (void) log_oom();
+
+                fprintf(f, "%sBPFDelegateCommands: %s\n", prefix, commands);
+                fprintf(f, "%sBPFDelegateMaps: %s\n", prefix, maps);
+                fprintf(f, "%sBPFDelegatePrograms: %s\n", prefix, programs);
+                fprintf(f, "%sBPFDelegateAttachments: %s\n", prefix, attachments);
```
Then, if _to_string() failed, `commands` and friends may be NULL. So please wrap them with `strna()`.

yuwata commented on src/core/load-fragment.c:
```diff
@@ -6626,3 +6627,40 @@ int config_parse_protect_hostname(
         return 1;
 }
+
+int config_parse_bpf_delegate(
```
Please use `DEFINE_CONFIG_PARSE_PTR()` marco and define 4 conf parsers.

teknoraver replied:
And put 4 different parsers in load-fragment-gperf.gperf.in?
I thought that a single parser with a switch was more efficient.

yuwata commented on src/basic/generate-bpf-delegate-configs.py:
```diff
@@ -0,0 +1,69 @@
+import sys
+
+def print_usage_and_exit():
+    print(f'Usage: {sys.argv[0]} header <header>')
```
I thought this generates a self-contained header file, but it does not, and generates .inc file.
https://github.com/systemd/systemd/pull/36134/commits/98f2e99850eb01d8e7bb8f3c3f4dc79fb2f8e986#r2169349245
Then, I think `code` is better. Sorry.

yuwata commented on src/core/namespace.c:
```diff
@@ -3967,6 +3970,201 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+DEFINE_STRING_TABLE_LOOKUP(bpf_delegate_prog_type, uint64_t);
+DEFINE_STRING_TABLE_LOOKUP(bpf_delegate_attach_type, uint64_t);
+
+char* bpf_delegate_commands_to_string(uint64_t u) {
```
Maybe better to make this slightly generic, and take a function pointer to bpf_delegate_cmd_to_string() and friends. That greatly reduce the number of lines.

yuwata commented on src/core/namespace.c:
```diff
@@ -3967,6 +3970,201 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+                return strdup("any");
+
+        _cleanup_free_ char *buf = NULL;
+
```
Typically, we do not insert an empty line like that. Though, not a big matter.

yuwata commented on src/core/namespace.c:
```diff
@@ -3967,6 +3970,201 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+        return TAKE_PTR(buf);
+}
+
+int bpf_delegate_commands_from_string(const char *s, uint64_t *ret) {
```
Also this one. I'd like to make this take a function pointer, then we can merge the for parsers.

yuwata commented on src/core/namespace.c:
```diff
@@ -3967,6 +3970,201 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+        }
+
+        uint64_t mask = 0;
+
```
Here also. I'd like to drop the empty line. But not a matter.

teknoraver commented:
@yuwata I have addressed all comments and also finished the documentation generation.

I used a better IMHO approach in this version: I generate a single XML file with the strings for all four parameters and then use `xpointer` to include the relevant part in each section.
So we generate the whole XML at once in a single file, the meson change is small, and the generator script is simpler.

Also, I set the default value for the delegates to 0, otherwise just setting PrivateBPF=yes would mount a bpffs with all BPF capabilities.
I investigated a bit and that's the default kernel behavior, no permissions unless explicitly allowed.

teknoraver commented on src/core/dbus-execute.c:
```diff
@@ -1115,6 +1115,90 @@ static int property_get_unsigned_as_uint16(
+
+        assert(reply);
+
+        s = bpf_delegate_to_string(*u, bpf_delegate_cmd_to_string);
```
The four helpers bpf_delegate_cmd_to_string, bpf_delegate_map_type_to_string, etc.
are created with DEFINE_STRING_TABLE_LOOKUP() against the code generated in `bpf-delegate-configs.inc`
That code gets the enum names directly from the kernel header, which are:
* bpf_cmd
* ‎bpf_map_type
* bpf_prog_type
* bpf_attach_type
So to rename them I should add a translation map in the generator script and it's not wort it IMHO.

see https://github.com/torvalds/linux/blob/v6.15/include/uapi/linux/bpf.h#L925-L1124

yuwata replied:
Sorry, I cannot follow.
BTW, my first comment has typo: _from_string() -> _to_string(), and I am requesting the following:
```diff
-        s = bpf_delegate_to_string(*u, bpf_delegate_cmd_to_string);
+        s = bpf_delegate_commands_to_string(*u);
```

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5506,10 +5588,27 @@ int exec_invoke(
 
+        if (context->private_bpf != PRIVATE_BPF_NO) {
+                r = pidref_wait_for_terminate_and_check("(sd-bpffs)", &bpffs_pid, /* flags = */ 0);
+                TAKE_PIDREF(bpffs_pid);
```
This should be moved after the error handling and be replaced with `pidref_done()`, otherwise we'll leak the pidref

DaanDeMeyer commented on src/core/exec-invoke.c:
```diff
@@ -5506,10 +5588,27 @@ int exec_invoke(
+                        *exit_status = EXIT_BPF;
+                        if (read(bpffs_errno_pipe, &r, sizeof(r)) == sizeof(r))
+                                return log_debug_errno(r, "bpffs helper exited with error: %m");
+                        return -EIO;
```
Please debug log when we can't read the error as well. And use `errno` so we know which error `read()` failed with instead of `EIO`.

DaanDeMeyer commented on test/units/TEST-07-PID1.private-bpf.sh:
```diff
@@ -0,0 +1,21 @@
+        -p PrivateBPF=no \
+        grep -q '/sys/fs/bpf .* ro,' /proc/mounts
+
+# Check that with PrivateBPF=yes, the host bpffs is mounted with the delegate_* options
```
"without the delegate_* options"?

DaanDeMeyer commented on src/test/test-bpf-token.c:
```diff
@@ -0,0 +1,20 @@
+#if __LIBBPF_CURRENT_VERSION_GEQ(1, 5)
+        _cleanup_close_ int bpffs_fd = -EBADF, token_fd = -EBADF;
+
+        ASSERT_OK_ERRNO(bpffs_fd = open("/sys/fs/bpf", O_RDONLY));
+        ASSERT_OK(token_fd = bpf_token_create(bpffs_fd, NULL));
```
No need to use these helpers, use DEFINE_MAIN_FUNCTION() and then just return regular errors instead of using the assertion macros, then you don't have to exclude the coredumps either.

DaanDeMeyer reviewed (changes requested):
Didn't finish reviewing the last commit yet

teknoraver commented:
@yuwata @DaanDeMeyer in some previous push I discovered that fsmount() can be called as unprivileged, so I moved it from the privileged child to the late user code.

Today, I decided to fix the kernel selftest accordingly, and while doing it I noticed that now, a SCM_RIGHTS exchange become unneeded. So I removed it:
https://github.com/kernel-patches/bpf/commit/212ec92295673a362453f8ede36033b61c7983fa

I ported this change to systemd too, now we have a SCM_RIGHTS less and also save a fd.

bluca commented:
```
712s FAILED: man/systemd.exec.5 
712s /usr/bin/xsltproc -o man/systemd.exec.5 --nonet --xinclude --maxdepth 9000 --stringparam man.output.quietly 1 --stringparam funcsynopsis.style ansi --stringparam man.authors.section.enabled 0 --stringparam man.copyright.section.enabled 0 --stringparam systemd.version '258~devel' --path /tmp/autopkgtest.1KOswn/build.tEo/systemd/obj-x86_64-linux-gnu/man:/tmp/autopkgtest.1KOswn/build.tEo/systemd/man:/tmp/autopkgtest.1KOswn/build.tEo/systemd/obj-x86_64-linux-gnu/src/shared ../man/custom-man.xsl ../man/systemd.exec.xml
712s warning: failed to load external entity "../man/bpf-delegate.xml"
712s ../man/systemd.exec.xml:2533: element include: XInclude error : could not load ../man/bpf-delegate.xml, and no fallback was found
712s warning: failed to load external entity "../man/bpf-delegate.xml"
712s ../man/systemd.exec.xml:2547: element include: XInclude error : could not load ../man/bpf-delegate.xml, and no fallback was found
712s warning: failed to load external entity "../man/bpf-delegate.xml"
712s ../man/systemd.exec.xml:2559: element include: XInclude error : could not load ../man/bpf-delegate.xml, and no fallback was found
712s warning: failed to load external entity "../man/bpf-delegate.xml"
712s ../man/systemd.exec.xml:2571: element include: XInclude error : could not load ../man/bpf-delegate.xml, and no fallback was found
```

yuwata reviewed (changes requested):
Sorry, please rebase.

teknoraver commented:
@yuwata yes I noticed the typo, but I changed it only in src/core/execute.c, I did it in src/core/dbus-execute.c too just now

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2223,6 +2223,53 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+        /* Disable the setgroups() system call in the child user namespace, for good, unless PrivateUsers=full
+         * and using the system service manager. */
+        const char *a = procfs_file_alloca(ppid, "setgroups");
+        _cleanup_close_ int fd = open(a, O_WRONLY|O_CLOEXEC);
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to open %s: %m", a);
+
+        const char *setgroups = allow_setgroups ? "allow\n" : "deny\n";
+        if (write(fd, setgroups, strlen(setgroups)) < 0)
+                return log_debug_errno(errno, "Failed to write '%s' to %s: %m", setgroups, a);
```
write_string_file

teknoraver replied:
Yes, I just moved that code block so I tried to touch as little as possible

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2223,6 +2223,53 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+
+        /* First write the GID map */
+        a = procfs_file_alloca(ppid, "gid_map");
+        fd = open(a, O_WRONLY|O_CLOEXEC);
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to open %s: %m", a);
+
+        if (write(fd, gid_map, strlen(gid_map)) < 0)
+                return log_debug_errno(errno, "Failed to write GID map to %s: %m", a);
```
ditto

YHNdnzj commented on src/core/exec-invoke.c:
```diff
@@ -2223,6 +2223,53 @@ static int build_pass_environment(const ExecContext *c, char ***ret) {
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to open %s: %m", a);
+
+        if (write(fd, uid_map, strlen(uid_map)) < 0)
```
ditto

teknoraver commented:
@yuwata @YHNdnzj hi, I think that the PR is code complete, can I get a final review?

yuwata commented on src/core/namespace.c:
```diff
@@ -1697,6 +1732,35 @@ static int mount_overlay(const MountEntry *m) {
+        if (r < 0)
+                return log_debug_errno(fs_fd, "Failed to receive data from child: %m");
+
+        _cleanup_close_ int mnt_fd = fsmount(fs_fd, 0, 0);
```
Please add short comment for the two `0`s.

yuwata commented on src/core/namespace.c:
```diff
@@ -3888,6 +3960,13 @@ static const char* const proc_subset_table[_PROC_SUBSET_MAX] = {
+        [PRIVATE_BPF_YES]   = "yes",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(private_bpf, PrivateBPF);
```
Please use `DEFINE_STRING_TABLE_LOOKUP_WITH_BOOLEAN()` to accept e.g. `true/false`, `1/0`.

yuwata commented on src/core/exec-invoke.c:
```diff
@@ -2259,6 +2259,62 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+
+                r = write(socket_fds[1], (char[]) { 0 }, sizeof(char));
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to send data to child: %m");
```
`r` -> `errno`

yuwata reviewed (approved):
Looks mostly good.
Several nitpicks.

teknoraver commented on src/core/namespace.c:
```diff
@@ -3967,6 +3970,71 @@ static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+        if (u == UINT64_MAX)
+                return strdup("any");
+
+        _cleanup_free_ char *buf = strdup("");
+        if (!buf)
+                return NULL;
```
I think it's more practical to keep the one in the beginnint and remove the one on return

yuwata reviewed (commented):
Almost there! Several more nitpicks.

poettering commented on src/nspawn/nspawn-mount.c:
```diff
@@ -516,6 +516,14 @@ int mount_sysfs(const char *dest, MountSettingsMask mount_settings) {
+                return log_oom();
+
+        (void) mkdir_p(x, 0755);
+
```
maybe use:
```c
FOREACH_STRING(p, "/fs/cgroup", "/fs/bpf"/) {
        _cleanup_free char *x = path_join(top, p);
        if (!x)
                return log_oom();
        (void) mkdir_p(x, 0755)
}
```

poettering commented on man/systemd.exec.xml:
```diff
@@ -2555,6 +2555,16 @@ RestrictNamespaces=~cgroup net</programlisting>
+        <term><varname>PrivateBPF=</varname></term>
+
+        <listitem><para>Takes a boolean argument. If set, mount a private instance of the BPF filesystem
+        on <filename>/sys/fs/bpf</filename>. Otherwise, if <varname>ProtectKernelTunables=</varname> is set,
```
plese suffix paths that refer to dirs with `/`, in accodance to our coding style, i.e. spell out `/sys/fs/bpf/` here.

poettering commented on src/core/exec-invoke.c:
```diff
@@ -5724,10 +5809,27 @@ int exec_invoke(
+                        *exit_status = EXIT_BPF;
+                        if (read(bpffs_errno_pipe, &r, sizeof(r)) == sizeof(r))
+                                return log_debug_errno(r, "bpffs helper exited with error: %m");
+                        return log_debug_errno(errno, "Failed to read from the bpffs helper errno pipe: %m");
```
read() might return short, who knows, and you'll output errno here, which is not initialized in case of a short read

poettering commented on src/core/namespace.c:
```diff
@@ -1697,6 +1732,35 @@ static int mount_overlay(const MountEntry *m) {
+        if (r < 0)
+                return log_debug_errno(r, "Failed to send bpffs fd to child: %m");
+
+        r = read(socket_fd, (uint8_t[1]) {}, 1);
```
as above, ssize_t vs int

poettering commented on src/core/namespace.c:
```diff
@@ -3888,6 +3960,13 @@ static const char* const proc_subset_table[_PROC_SUBSET_MAX] = {
 
+static const char* const private_bpf_table[_PRIVATE_BPF_MAX] = {
+        [PRIVATE_BPF_NO]    = "no",
+        [PRIVATE_BPF_YES]   = "yes",
```
aligning the `=` is great, but please remove redundant whitespace

yuwata commented:
@teknoraver Please double check about the short comments for function arguments. Still missing `=` at several places.

teknoraver commented:
@yuwata what is the rule behind sort comments?
i.e. when they are needed, when the argument is 0/NULL?

teknoraver commented on src/core/exec-invoke.c:
```diff
@@ -2270,6 +2270,62 @@ static int setup_private_users_child(int unshare_ready_fd, const char *uid_map,
+                        report_errno_and_exit(bpffs_errno_pipe[1], errno);
+                }
+
+                r = write(socket_fds[1], (uint8_t[1]) {}, 1);
```
what about:
```c
if (write(socket_fds[1], (uint8_t[1]) {}, 1) < 0)
...
```

teknoraver commented:
@bluca the mkosi/ci failures on this run were on x86_64 and unrelated with this change.
Where can I get an arm64 or ppc64 log?
